schema.go

```diff
@@ -0,0 +1,149 @@
+// like resources, data sources, and providers. Think of it as a type
+// definition, but for Terraform.
+type Schema struct {
+	// Attributes are the fields inside the resource, provider, or data
+	// source that the schema is defining. The map key should be the name
+	// of the attribute, and the body defines how it behaves. Names must
+	// only contain lowercase letters, numbers, and underscores.
+	Attributes map[string]Attribute
+
+	// Version indicates the current version of the schema. Schemas are
+	// versioned to help with automatic upgrade process. Whenever you have
+	// a change in the schema you'd like to provide a manual migration for,
+	// you should increment that schema's version by one.
+	Version int64
+}
+
+// Attribute defines the constraints and behaviors of a single field in a
+// schema. Attributes are the fields that show up in Terraform state files and
+// can be used in configuration files.
+type Attribute struct {
+	// Type indicates what kind of attribute this is. You'll most likely
+	// want to use one of the types in the types package.
+	//
+	// If Type is set, Attributes cannot be.
+	Type AttributeType
+
+	// Attributes can have their own, nested attributes. This nested map of
+	// attributes behaves exactly like the map of attributes on the Schema
+	// type.
+	//
+	// If Attributes is set, Type cannot be.
+	//
+	// TODO: do we need MaxItems/MinItems? Can we just make those weird
+	// validation helpers?
+	Attributes map[string]Attribute
+
+	// AttributesNestingMode controls the various ways these sub-groups of
+	// attributes can behave. It can only be used with Attributes, and must
+	// not be set if Type is set.
+	AttributesNestingMode NestingMode
+
+	// Description is used in various tooling, like the documentation
+	// generator and the language server, to give practitioners more
+	// information about what this attribute is, what it's for, and how it
+	// should be used.
+	Description string
+
+	// DescriptionKind indicates the type of text formatting that
+	// Description uses. It should be Markdown or PlainText.
+	//
+	// TODO: come up with a better interface for this, this is weird.
+	DescriptionKind StringKind
+
+	// Required indicates whether the practitioner must enter a value for
+	// this attribute or not. Required and Optional cannot both be true,
+	// and Required and Computed cannot both be true.
+	Required bool
+
+	// Optional indicates whether the practitioner can choose not to enter
+	// a value for this attribute or not. Optional and Required cannot both
+	// be true.
+	Optional bool
+
+	// Computed indicates whether the provider may return its own value for
+	// this attribute or not. Required and Computed cannot both be true. If
+	// Required and Optional are both false, Computed must be true, and the
+	// attribute will be considered "read only" for the practitioner, with
+	// only the provider able to set its value.
+	Computed bool
```

Comment:
Passing thought: Is now a good time to introduce this as a single enumeration type instead of allowing plugin developers potentially introduce invalid combinations and allowing for easier static analysis and unit testing (this one field being required)?

Quick sketch being:

```go
const (
    AttributeConfigurationRequired AttributeConfiguration = 0
    AttributeConfigurationOptional AttributeConfiguration = 1
    AttributeConfigurationReadOnly AttributeConfiguration = 2
)

type AttributeConfiguration uint8

type Attribute struct {
    // ... other fields ...
    Configuration AttributeConfiguration
}
```

Reply:
Now's a great time if it's something we want to introduce! I'm just not convinced it's a thing we want to introduce. I think it was @apparentlymart who attempted to combine required/optional/computed (maybe not in the way you're suggesting) and determined he needed all three booleans to support all the nuance.

I'm interested in hearing more how this impacts static analysis or unit testing.

My concern with the example you shared:

* The distance between where the options are enumerated and their effect explained, and where they're used
* The verbosity that hosting them to global scope forces on us (I don't think anyone actually wants to type `AttributeConfigurationOptionalWithMissingDriftDetection` instead of `Optional: true`.)
* I think this framing is reductive about what `Computed` does. `AttributeConfigurationOptionalWithoutMissingDriftDetection` is (assuming that's Optional + Computed) actually also going to be used for, e.g., providing a default and may have nothing to do with drift detection.

I like the idea of trying to enforce only valid configurations with the compiler, but I'm not sure what we lose in this scenario is worth the benefit. Thought hearing more about the static analysis and unit testing impacts may change my mind.

Reply:
I'm thinking more and more about this, and I think it's a useful and valid direction for us to pursue, but I'm not sold on it just yet. I think my hot take is that we should create an issue for exploring this, and schedule that work for the v0.2.0 release.

Reply:
I have created https://github.com/hashicorp/terraform-plugin-framework/issues/31 to track this investigation longer term. 👍

Reply:
Thank you!